pydantic_ai_slim/pydantic_ai/ext/temporal/_mcp_server.py

```diff
@@ -0,0 +1,119 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any, Callable, Literal
+
+from pydantic import ConfigDict, with_config
+from temporalio import activity, workflow
+from temporalio.workflow import ActivityConfig
+
+from pydantic_ai._run_context import RunContext
+from pydantic_ai.exceptions import UserError
+from pydantic_ai.mcp import MCPServer, ToolResult
+from pydantic_ai.tools import ToolDefinition
+from pydantic_ai.toolsets.abstract import ToolsetTool
+
+from ._run_context import TemporalRunContext
+from ._toolset import TemporalWrapperToolset
+
+
+@dataclass
+@with_config(ConfigDict(arbitrary_types_allowed=True))
+class _GetToolsParams:
+    serialized_run_context: Any
+
+
+@dataclass
+@with_config(ConfigDict(arbitrary_types_allowed=True))
+class _CallToolParams:
+    name: str
+    tool_args: dict[str, Any]
+    serialized_run_context: Any
+    tool_def: ToolDefinition
+
+
+class TemporalMCPServer(TemporalWrapperToolset):
+    def __init__(
+        self,
+        server: MCPServer,
+        activity_config: ActivityConfig = {},
+        tool_activity_config: dict[str, ActivityConfig | Literal[False]] = {},
+        run_context_type: type[TemporalRunContext] = TemporalRunContext,
+    ):
+        super().__init__(server)
+        self.activity_config = activity_config
+        self.tool_activity_config = tool_activity_config
+        self.run_context_type = run_context_type
+
+        id = server.id
+        assert id is not None
```

Comment:
Probably want an actual error message here

Reply:
@dmontagu We have it in `TemporalAgent`, by the time we get here we know they have IDs.